### PR TITLE
emacsWrapper: Put autoloads behind optional flag

### PR DESCRIPTION
**Pull request by nagy**
###### Description of changes

This puts the recently introduced autoloads for emacs behind an optional
flag, which is off by default.

See https://github.com/NixOS/nixpkgs/pull/168954 for more information.

@myguidingstar can you please check if this works for you?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by myguidingstar** [comment]
I've looked at the code and it looks great, but I don't have the computer to really run it soon

File: pkgs/build-support/emacs/wrapper.nix [hunk 1]
```diff
@@ -228,7 +232,8 @@ runCommand
         --subst-var-by bash ${emacs.stdenv.shell} \
         --subst-var-by wrapperSiteLisp "$deps/share/emacs/site-lisp" \
         --subst-var-by wrapperSiteLispNative "$deps/share/emacs/native-lisp:" \
-        --subst-var-by autoloadExpression "-l cl-loaddefs -l nix-generated-autoload" \
+        --subst-var-by autoloadExpression \
+          "${if withAutoloads then "'-l cl-loaddefs -l nix-generated-autoload'" else "''"}" \
```
Comment (AndersonTorres) [inline pkgs/build-support/emacs/wrapper.nix:236]: `lib.optionalString` would be more interesting

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/build-support/emacs/wrapper.nix (hunk 1 above)
Comment (nagy) [reply pkgs/build-support/emacs/wrapper.nix]: Updated, but I could not test it because I dont have a Darwin machine.

**Review by AndersonTorres — APPROVED** [review]

**Comment by adisbladis** [comment]
I can't say I like the idea of having flags to work around bugs in our wrapper code.. I think we should either disable the functionality outright or fix it.
Not add more flags.

**Comment by nagy** [comment]
The added autoload code for this wrapper has helped me to reduce my configs a good bit because I can now use functions without having to know what package they are coming from and `require`ing them. This is also very helpful if you want to call emacs functions from the CLI. But this code can cause problems when there is a badly misbehaving package like what arose here https://github.com/NixOS/nixpkgs/pull/168954#issuecomment-1114806352 , when though I made sure to call the `#'load` function with the `NO-ERROR` argument.

> I think we should either disable the functionality outright or fix it.

I am fine with both. I think that others can also benefit from this autoloading but, since I cannot reproduce the error, I can not work on any fix. In which way should we proceed?

**Comment by nagy** [comment]
I have attempted to fix the underlying issue, with the hope that we can get rid of that flag and close this PR. I attempted to build a wrapper with more and more packages from `emacs.pkgs` and I saw more and more errors, meaning that not all packages are able to being autoloaded. However, when wrapping the generated `(load ...)` calls with `ignore-errors` everything worked fine.

My code for this is here:
Commit: https://github.com/nagy/nixpkgs/commit/f854fccb2d2905c8a6bd952739f82e079f52f0c1
Branch: https://github.com/nagy/nixpkgs/tree/emacs-ignore-autoload-errors

@myguidingstar could you verify that this fix would solve the issue for you? If so, then we can close this PR here and I can make another one for that fix.

**Comment by nagy** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/181174

**Pull request closed without merging**